pkg/cmd/infra/router/template.go

```diff
@@ -287,6 +293,83 @@ func parseCaptureHeaders(in string) ([]templateplugin.CaptureHTTPHeader, error)
 	return captureHeaders, nil
 }
 
+func sanitizeHeaderValue(headerValue string) string {
+	sanitizeHeaderValue := strings.ReplaceAll(headerValue, "'", `'\''`)
+	return "'" + sanitizeHeaderValue + "'"
+}
+
+func parseHeadersToBeSetOrDeleted(in string) ([]templateplugin.HTTPHeader, error) {
+	var captureHeaders []templateplugin.HTTPHeader
+	var capture templateplugin.HTTPHeader
+	var err error
+	if len(in) == 0 {
+		return captureHeaders, fmt.Errorf("encoded header string not present.")
+	}
+	if len(in) > 0 {
+		for _, header := range strings.Split(in, ",") {
+			parts := strings.Split(header, ":")
+			num := len(parts)
+			switch num {
+			default:
+				{
+					return captureHeaders, fmt.Errorf("invalid HTTP header input specification: %v", header)
+				}
+			case 3:
+				{
+					headerName, err := url.QueryUnescape(parts[0])
+					if err != nil {
+						return captureHeaders, fmt.Errorf("failed to decode percent encoding: %v", parts[0])
+					}
+					// RFC 2616, section 4.2, states that the header name
+					// must be a valid token.
+					if !validTokenRE.MatchString(headerName) {
+						return captureHeaders, fmt.Errorf("invalid HTTP header name: %v", headerName)
+					}
+					headerValue, err := url.QueryUnescape(parts[1])
+					if err != nil {
+						return captureHeaders, fmt.Errorf("failed to decode percent encoding: %v", parts[1])
+					}
+					sanitizedHeaderValue := templateplugin.SanitizeHeaderValue(headerValue)
+					action, err := url.QueryUnescape(parts[2])
+					if err != nil {
+						return captureHeaders, fmt.Errorf("failed to decode percent encoding: %v", parts[2])
+					}
+					capture = templateplugin.HTTPHeader{
+						Name:   headerName,
+						Value:  sanitizedHeaderValue,
+						Action: routev1.RouteHTTPHeaderActionType(action),
+					}
+					captureHeaders = append(captureHeaders, capture)
+				}
+			case 2:
+				{
+					headerName, err := url.QueryUnescape(parts[0])
+					if err != nil {
+						return captureHeaders, fmt.Errorf("failed to decode percent encoding: %v", parts[0])
+					}
+					// RFC 2616, section 4.2, states that the header name
+					// must be a valid token.
+					if !validTokenRE.MatchString(headerName) {
+						return captureHeaders, fmt.Errorf("invalid HTTP header name: %v", headerName)
+					}
```

Comment:
This code is called twice, why not refactor into a function?

Reply:
yes created functions now